mkdocs/config/config_options.py

```diff
@@ -0,0 +1,333 @@
+import os
+
+import six
+
+from mkdocs import utils, legacy
+from mkdocs.legacy import OrderedDict
+
+
+class ValidationError(Exception):
+    """Raised during the validation process of the config on errors."""
+
+
+class BaseConfigOption(object):
+    """
+    The BaseConfigOption adds support for default values and required values
+
+    It then delegates the validation and (optional) post processing to
+    subclasses.
+    """
+
+    def __init__(self, default=None, required=False):
+        self.default = default
+        self.required = required
+        self.warnings = []
+
+    def is_required(self):
+        return self.required
+
+    def validate(self, value):
+        """
+        Perform some initial validation.
+
+        If the option is empty (None) and isn't required, leave it as such. If
+        it is empty but has a default, use that. Finally, call the _validate
+        method on the subclass unless.
+        """
+
+        if value is None:
+            if self.default is not None:
+                value = self.default
+            elif not self.required:
+                return
+            elif self.required:
+                raise ValidationError("Required configuration not provided.")
+
+        return self._validate(value)
+
+    def _validate(self, value):
+        """
+        Perform validation for a value.
+
+        The _validate method should be implemented by subclasses.
```

Comment:
The way I see it, if it is all part of the same code base (not just the same class) then it is okay to call the protected method internally. However, the public facing API should never be private/protected. In other words, third party code should never call or override in a subclass any method which starts with an underscore (Python's special double underscore methods being the exception). However, with my suggested change, the protected method is only ever used by mkdocs internally (although, admittedly, not from within the same class). If it ever changes in the future, third party code (or subclasses) would never need to change. Who cares if the internal mkdocs code needs to change (not including subclasses as that gets repeated many times over). 

But if you want to use `do_validation` instead and avoid the underscore altogether, then that works too. Even then, I would name the method that gets overridden `validate` and use `do_validation` for the internal method (what I was suggesting you call `_validate`).

Oh and @facelessuser is correct, a single underscore generally indicates a protected method. - although that is not actually enforced by Python. It is just good practice. Double underscore methods are private which is enforced. Sorry I used the wrong word earlier.